packages/databricks-vscode/resources/python/notebook.workflow-wrapper.py

```diff
@@ -1,12 +1,54 @@
-#This file is autogenerated by the Databricks Extension for VS Code
+# This file is autogenerated by the Databricks Extension for VS Code
+def databricks_preamble():
+    from IPython import get_ipython
+    from typing import List
 
-dbutils.widgets.text("DATABRICKS_SOURCE_FILE", "")
-dbutils.widgets.text("DATABRICKS_PROJECT_ROOT", "")
+    dbutils.widgets.text("DATABRICKS_SOURCE_FILE", "")
+    dbutils.widgets.text("DATABRICKS_PROJECT_ROOT", "")
+    src_file_dir = None
+    project_root_dir = None
 
-if dbutils.widgets.get("DATABRICKS_SOURCE_FILE") != "":
-	import os
-	os.chdir(os.path.dirname(dbutils.widgets.get("DATABRICKS_SOURCE_FILE")))
+    if dbutils.widgets.get("DATABRICKS_SOURCE_FILE") != "":
+        import os
+        src_file_dir = os.path.dirname(
+            dbutils.widgets.get("DATABRICKS_SOURCE_FILE"))
+        os.chdir(src_file_dir)
 
-if dbutils.widgets.get("DATABRICKS_PROJECT_ROOT") != "":
-	import sys
-	sys.path.insert(0, dbutils.widgets.get("DATABRICKS_PROJECT_ROOT"))
+    if dbutils.widgets.get("DATABRICKS_PROJECT_ROOT") != "":
+        import sys
+        project_root_dir = dbutils.widgets.get("DATABRICKS_PROJECT_ROOT")
+        sys.path.insert(0, project_root_dir)
+
+    def parse_databricks_magic_lines(lines: List[str]):
+        if len(lines) == 0:
+            return lines
+
+        first = ""
+        for line in lines:
+            if len(line.strip()) != 0:
+                first = line
+                break
+
+        if first.startswith("%"):
+            magic = first.split(" ")[0].strip().strip("%")
+            rest = ' '.join(first.split(" ")[1:])
+
+            if magic == "sh":
+                return [
+                    "%sh\n",
+                    f"cd {src_file_dir}\n",
```

Comment:
the dir needs to be escaped in case the path contains spaces or special characters.

Reply:
Right. Good catch.